### ISSUE TITLE
Cannot inject dependency which is a subclass of a subscripted type

### ISSUE THREAD

**Issue by eugenhu** [alecthomas/injector#128]
Running the example code below,

```python
from typing import Sequence
from injector import inject, Injector


class FibonacciSequence(Sequence[int]):
    def __init__(self, n_terms: int = 10) -> None:
        self._n_terms = n_terms

    def __getitem__(self, i: int) -> int:
        if i >= self._n_terms: raise IndexError

        if i == 0: return 0
        if i == 1: return 1
        if i >= 2: return self[i-2] + self[i-1]

    def __len__(self) -> int:
        return self._n_terms


# Test injecting FibonacciSequence dependency

@inject
def main(sequence: FibonacciSequence) -> None:
    print(tuple(sequence))


injector = Injector()
injector.call_with_injection(main)
```

I get the following error,

```
Traceback (most recent call last):
  ...
  File ".../python3.5/site-packages/injector/__init__.py", line 1072, in _infer_injected_bindings
    v, metadata = v.__origin__, v.__metadata__
AttributeError: type object 'FibonacciSequence' has no attribute '__metadata__'
```

I ran this on Python 3.5.7.

I think the cause of this issue is `_is_specialization(v, Annotated)` incorrectly returns `True` when `v` is `FibonacciSequence`.

https://github.com/alecthomas/injector/blob/a461f722adcf8e23f0fdd4e122f101029a74877c/injector/__init__.py#L1071-L1073

The current implementation of `_is_specialization()` is

https://github.com/alecthomas/injector/blob/a461f722adcf8e23f0fdd4e122f101029a74877c/injector/__init__.py#L596-L618

The value of `FibonacciSequence.__origin__` is `None`. I'm not sure where exactly the `__origin__` attribute is set but it seems (I'm guessing) that it is set to `None` for unsubscripted generics or subclasses of subscripted generics. So the `not hasattr(cls, '__origin__')` condition in line 608 fails as `FibonacciSequence` has an `__origin__` attribute set to `None`. Finally on line 618, the comparison `origin is generic_class` succeeds because here `origin` is `None`, and `generic_class` is also `None` because I am running this on a version of Python lower than 3.7.0 so `HAVE_ANNOTATED` is `False` and so `Annotated` has also been set to `None`.

I think modifying line 608 to
```python
if getattr(cls, '__origin__', None) is None:
```
should fix this issue.

**Comment by jstasiak** [comment]
Hey, a fix has just landed in the master branch. The patch is different than yours as I needed to resolve a new issue discovered by new mypy (0.750) anyway but this doesn't take away from your analysis of the bug – it was thorough and detailed, an exemplary report in my book. Thank you!

**Comment by jstasiak** [comment]
The fix for this has just been released as part of the version 0.18.1.
